### DISCUSSION TITLE
Limit Workout Downloads to Most Recent

### DISCUSSION THREAD

**Discussion by sarpete309** [philosowaffle/peloton-to-garmin#214] [Q&A]
I've searched for a solution but couldn't find anything. How can I limit the downloads to the most recent 2-3 workouts? Every time I run the program, it pulls every workout I've ever done. Thanks!!

**Comment by philosowaffle** [comment]
In the [Peloton Config](https://philosowaffle.github.io/peloton-to-garmin/configuration/json.html#peloton-config) section there is a setting for `NumWorkoutsToDownload`. Set this to a number, say 3, and it will download the 3 most recent workouts.

By default, if you don't have this value set, P2G should attempt to ask you to enter a number, or fallback to not downloading any workouts at all.

**Reply by sarpete309** [reply]
I currently have it set to 2, but it still downloads all workouts I've completed. 

Here's my config file (minus my acutal login info):
{
  "App": {
    "OutputDirectory": "./output",
    "WorkingDirectory": "./working",
    "SyncHistoryDbPath": "./output/syncHistory.json",
    "EnablePolling": true,
    "PollingIntervalSeconds": 86400,
    "CloseWindowOnFinish": false
  },

  "Format": {
    "Fit": true,
    "Json": false,
    "Tcx": false,
    "SaveLocalCopy": true,
    "IncludeTimeInHRZones": false,
    "IncludeTimeInPowerZones": false,
    "DeviceInfoPath": "./deviceInfo.xml"
  },

  "Peloton": {
    "Email": "",
    "Password": "",
    "NumWorkoutsToDownload": 2,
    "ExcludeWorkoutTypes": [ "meditation" ]
  },

  "Garmin": {
    "Email": "",
    "Password": "",
    "Upload": true,
    "FormatToUpload": "fit",
    "UploadStrategy": 2
  },

  "Observability": {

    "Prometheus": {
      "Enabled": false,
      "Port": 4000
    },

    "Jaeger": {
      "Enabled": false,
      "AgentHost": "localhost",
      "AgentPort": 6831
    },

    "Serilog": {
      "Using": [ "Serilog.Sinks.Console", "Serilog.Sinks.File" ],
      "MinimumLevel": "Information",
      "WriteTo": [
        { "Name": "Console" },
        {
          "Name": "File",
          "Args": {
            "path": "./output/log.txt",
            "rollingInterval": "Day",
            "retainedFileCountLimit": 7
          }
        }
      ]
    }
  }
}

**Reply by philosowaffle** [reply]
Well that certainly is odd. Your config file looks all correct.  Do you mind changing this line `"MinimumLevel": "Information",` to be `"MinimumLevel": "Debug",` and running it again? And you if you could share the logfile output please.  I'm not able to reproduce on my end so just trying to figure out whats different.

**Reply by sarpete309** [reply]
Here's the log file
[log20220204.txt](https://github.com/philosowaffle/peloton-to-garmin/files/8002738/log20220204.txt)
.

**Reply by philosowaffle** [reply]
Good news, its not downloading all your workouts from Peloton each time.
```
2022-02-04 06:47:24.376 -05:00 [DBG] Total workouts found: 2
2022-02-04 06:47:24.379 -05:00 [DBG] Total workouts found after filtering InProgress and ExcludeWorkoutTypes: 2
2022-02-04 06:47:24.419 -05:00 [DBG] Workout e479ff7e9e8b4faa97f9a363d2a91ff8 already downloaded, skipping.
2022-02-04 06:47:24.422 -05:00 [DBG] Workout a19a69ed363d4720af3def9b5bc04909 already downloaded, skipping.
```

But I can see why it looks that way, and something is still not working as intended based on the log output.  I noticed at the end that it's failing to authenticate with Garmin:
```
2022-02-04 06:47:26.929 -05:00 [ERR] Authentication Failed.
Flurl.Http.FlurlHttpException: Call failed with status code 403 (Forbidden): POST https://sso.garmin.com
```

Now in this scenario, where we fail to upload to Garmin, P2G intentionally saves the workout files locally with the goal to retry on the next run assuming it can communicate with Garmin again.  That seems to be what's happening here, purely based on the logs.

Next two questions:
1. Are the workouts mentioned in the log file all already uploaded on Garmin Connect, or have they never successfully made it there?
2. Triple check your Garmin credentials are correct in the config file.  If your password has any `\` or `'` in it, they [may need to be escaped](https://github.com/philosowaffle/peloton-to-garmin/discussions/163) so that they are read correctly by P2G. There is an online [validator tool](https://jsonlint.com/?code=) you can use to help determine if this is the problem.

**Comment by danielsfrazier** [comment]
Looks like maybe I am hitting a similar problem after updating to the
latest version today. Last time I updated was some time back in December
and it was working fine till I updated today.

I see the uploaded files in the syncHistory.json file but they don't clear
out of the working folder after they have been uploaded. I verified they
exist in the fit folder too. Looks like they try to upload the same file
every time I restart the program. I made sure it wasn't a permission
problem deleting files from the working dir or subdirs.

116
./downloaded/901f3254bf124ad6aee621897995dfd0_20_min_AT's_Ride_to_Greatness-_Game_1_with_Alex_Toussaint.json
80
 ./downloaded/c724a524789042f6bed78da82c8acfdf_5_min_AT's_Ride_to_Greatness-_Warm_Up_with_Alex_Toussaint.json
136
./downloaded/a623f514b84c4b44a050d1c74f7b6a6f_10_min_Warm_Up_Ride_with_Leanne_Hainsby.json
76
 ./downloaded/e4e3096c83e44dc9bfeef79e0e0e74d1_5_min_AT's_Ride_to_Greatness-_Cool_Down_with_Alex_Toussaint.json
412     ./downloaded
16
 ./upload/901f3254bf124ad6aee621897995dfd0_20_min_AT's_Ride_to_Greatness-_Game_1_with_Alex_Toussaint.Fit
8
./upload/c724a524789042f6bed78da82c8acfdf_5_min_AT's_Ride_to_Greatness-_Warm_Up_with_Alex_Toussaint.Fit
8
./upload/a623f514b84c4b44a050d1c74f7b6a6f_10_min_Warm_Up_Ride_with_Leanne_Hainsby.Fit
8
./upload/e4e3096c83e44dc9bfeef79e0e0e74d1_5_min_AT's_Ride_to_Greatness-_Cool_Down_with_Alex_Toussaint.Fit
44      ./upload
16
 ./901f3254bf124ad6aee621897995dfd0_20_min_AT's_Ride_to_Greatness-_Game_1_with_Alex_Toussaint.Fit
8
./c724a524789042f6bed78da82c8acfdf_5_min_AT's_Ride_to_Greatness-_Warm_Up_with_Alex_Toussaint.Fit
8
./a623f514b84c4b44a050d1c74f7b6a6f_10_min_Warm_Up_Ride_with_Leanne_Hainsby.Fit
8
./e4e3096c83e44dc9bfeef79e0e0e74d1_5_min_AT's_Ride_to_Greatness-_Cool_Down_with_Alex_Toussaint.Fit
500     ./

***@***.*** peloton-to-garmin]$ dotnet run --project
./src/PelotonToGarminConsole/PelotonToGarminConsole.csproj
Welcome! P2G is starting up...
[20:04:30 DBG] Hosting starting
[20:04:30 DBG] App Version: 2.1.1
[20:04:30 DBG] Operating System: Unix 5.4.17.2136
[20:04:30 DBG] DotNet Runtime: 5.0.13
[20:04:30 INF] Application started. Press Ctrl+C to shut down.
[20:04:30 INF] Hosting environment: Production
[20:04:30 INF] Content root path:
/home/myuser/peloton-to-garmin/src/PelotonToGarminConsole
[20:04:30 DBG] Hosting started
[20:04:31 DBG] Fetching recent workouts page: 0
[20:04:31 DBG] Total workouts found: 10
[20:04:31 DBG] Total workouts found after filtering InProgress and
ExcludeWorkoutTypes: 10
[20:04:31 DBG] Workout e4e3096c83e44dc9bfeef79e0e0e74d1 already downloaded,
skipping.
[20:04:31 DBG] Workout 901f3254bf124ad6aee621897995dfd0 already downloaded,
skipping.
[20:04:31 DBG] Workout c724a524789042f6bed78da82c8acfdf already downloaded,
skipping.
[20:04:31 DBG] Workout a623f514b84c4b44a050d1c74f7b6a6f already downloaded,
skipping.
[20:04:31 DBG] Workout e81f88afa01a4dd789684848395df522 already downloaded,
skipping.
[20:04:31 DBG] Workout 7834f58b7a11411b9a435cd360472a06 already downloaded,
skipping.
[20:04:31 DBG] Workout c9b043ce209f41bb942865a7035aa8f2 already downloaded,
skipping.
[20:04:31 DBG] Workout d76812495a7f46d8b1d7b41a5c538abb already downloaded,
skipping.
[20:04:31 DBG] Workout 0573fef365624e20b2e43d6f5a645090 already downloaded,
skipping.
[20:04:31 DBG] Workout a5cd46fe247543d2aa6d200922e3f7cd already downloaded,
skipping.
[20:04:32 DBG] No incline found.
[20:04:32 DBG] No incline found.
[20:04:32 DBG] No incline found.
[20:04:32 INF] Encoded FIT file
/home/myuser/peloton-to-garmin/src/PelotonToGarminConsole/working/901f3254bf124ad6aee621897995dfd0_20_min_AT's_Ride_to_Greatness-_Game_1_with_Alex_Toussaint.Fit
[20:04:32 INF] Backed up file
./output/fit/901f3254bf124ad6aee621897995dfd0_20_min_AT's_Ride_to_Greatness-_Game_1_with_Alex_Toussaint.Fit
[20:04:32 DBG] Prepped Fit file
./working/upload/901f3254bf124ad6aee621897995dfd0_20_min_AT's_Ride_to_Greatness-_Game_1_with_Alex_Toussaint.Fit
for upload.
[20:04:32 DBG] No incline found.
[20:04:32 DBG] No incline found.
[20:04:32 DBG] No incline found.
[20:04:32 INF] Encoded FIT file
/home/myuser/peloton-to-garmin/src/PelotonToGarminConsole/working/c724a524789042f6bed78da82c8acfdf_5_min_AT's_Ride_to_Greatness-_Warm_Up_with_Alex_Toussaint.Fit
[20:04:32 INF] Backed up file
./output/fit/c724a524789042f6bed78da82c8acfdf_5_min_AT's_Ride_to_Greatness-_Warm_Up_with_Alex_Toussaint.Fit
[20:04:32 DBG] Prepped Fit file
./working/upload/c724a524789042f6bed78da82c8acfdf_5_min_AT's_Ride_to_Greatness-_Warm_Up_with_Alex_Toussaint.Fit
for upload.
[20:04:32 DBG] No incline found.
[20:04:32 DBG] No incline found.
[20:04:32 DBG] No incline found.
[20:04:32 INF] Encoded FIT file
/home/myuser/peloton-to-garmin/src/PelotonToGarminConsole/working/a623f514b84c4b44a050d1c74f7b6a6f_10_min_Warm_Up_Ride_with_Leanne_Hainsby.Fit
[20:04:32 INF] Backed up file
./output/fit/a623f514b84c4b44a050d1c74f7b6a6f_10_min_Warm_Up_Ride_with_Leanne_Hainsby.Fit
[20:04:32 DBG] Prepped Fit file
./working/upload/a623f514b84c4b44a050d1c74f7b6a6f_10_min_Warm_Up_Ride_with_Leanne_Hainsby.Fit
for upload.
[20:04:32 DBG] No incline found.
[20:04:32 DBG] No incline found.
[20:04:32 DBG] No incline found.
[20:04:32 INF] Encoded FIT file
/home/myuser/peloton-to-garmin/src/PelotonToGarminConsole/working/e4e3096c83e44dc9bfeef79e0e0e74d1_5_min_AT's_Ride_to_Greatness-_Cool_Down_with_Alex_Toussaint.Fit
[20:04:32 INF] Backed up file
./output/fit/e4e3096c83e44dc9bfeef79e0e0e74d1_5_min_AT's_Ride_to_Greatness-_Cool_Down_with_Alex_Toussaint.Fit
[20:04:32 DBG] Prepped Fit file
./working/upload/e4e3096c83e44dc9bfeef79e0e0e74d1_5_min_AT's_Ride_to_Greatness-_Cool_Down_with_Alex_Toussaint.Fit
for upload.
[20:04:32 INF] Beginning Garmin Upload.
[20:04:32 INF] Uploading to Garmin with the following parameters: gupload
-u **email** -p **password**
"./working/upload/901f3254bf124ad6aee621897995dfd0_20_min_AT's_Ride_to_Greatness-_Game_1_with_Alex_Toussaint.Fit"
"./working/upload/c724a524789042f6bed78da82c8acfdf_5_min_AT's_Ride_to_Greatness-_Warm_Up_with_Alex_Toussaint.Fit"
"./working/upload/a623f514b84c4b44a050d1c74f7b6a6f_10_min_Warm_Up_Ride_with_Leanne_Hainsby.Fit"
"./working/upload/e4e3096c83e44dc9bfeef79e0e0e74d1_5_min_AT's_Ride_to_Greatness-_Cool_Down_with_Alex_Toussaint.Fit"
[20:04:36 INF] DEPRECATION: The OpenSSL being used by this python install
(OpenSSL 1.0.2k-fips  26 Jan 2017) does not meet the minimum supported
version (>= OpenSSL 1.1.1) in order to support TLS 1.3 required by
Cloudflare, You may encounter an unexpected Captcha or cloudflare 1020
blocks.

[20:04:36 INF] GUpload: 2022-02-03 20:04:32,921 [INFO] Try to login on
GarminConnect...
2022-02-03 20:04:35,417 [INFO] Logged in as myuser
2022-02-03 20:04:35,584 [INFO] Activity already uploaded 8233927839
<http://voice.google.com/calls?a=nc,%2B18233927839>:
901f3254bf124ad6aee621897995dfd0_20_min_AT's_Ride_to_Greatness-_Game_1_with_Alex_Toussaint.Fit
2022-02-03 20:04:35,715 [INFO] Activity already uploaded 8233927861 :
c724a524789042f6bed78da82c8acfdf_5_min_AT's_Ride_to_Greatness-_Warm_Up_with_Alex_Toussaint.Fit
2022-02-03 20:04:35,882 [INFO] Activity already uploaded 8233927883 :
a623f514b84c4b44a050d1c74f7b6a6f_10_min_Warm_Up_Ride_with_Leanne_Hainsby.Fit
2022-02-03 20:04:36,021 [INFO] Activity already uploaded 8234042767 :
e4e3096c83e44dc9bfeef79e0e0e74d1_5_min_AT's_Ride_to_Greatness-_Cool_Down_with_Alex_Toussaint.Fit
2022-02-03 20:04:36,021 [INFO] All done.

**Reply by philosowaffle** [reply]
Interesting. Ok so in your case it did not download any new workouts.  But because there were left over workouts from a previous run still in the download directory, it proceeded to re-convert those and attempt to upload to Garmin. This is identical to what was mentioned over on #211 .


I may not be able to investigate more till the weekend.  

In the meantime, it is safe to manually delete any files lingering in respective output, working, or downloaded directories, which should prevent the duplicate upload attempts.  Or roll back to the [stable version](https://github.com/philosowaffle/peloton-to-garmin/releases/tag/v2.1.0)

Issue #220
